Error while checking build/O2/fullCI_slc9 for 855ea9c14b2a72baa8dc76676c32700e0ae8e630 at 2025-06-10 15:53:
```
## sw/BUILD/O2-latest/log
/sw/SOURCES/O2/14379-slc9_x86-64/0/Detectors/TPC/workflow/src/ClusterDecoderRawSpec.cxx:89:20: error: no matching function for call to 'swap(o2::header::Stack&, o2::header::Stack&)'
/sw/slc9_x86-64/GCC-Toolchain/v14.2.0-alice2-1/include/c++/14.2.0/type_traits:138:11: error: no type named 'type' in 'struct std::enable_if<false, void>'
/sw/SOURCES/O2/14379-slc9_x86-64/0/Detectors/TPC/workflow/src/ClusterDecoderRawSpec.cxx:98:18: error: no matching function for call to 'swap(o2::header::Stack&, o2::header::Stack&)'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14379/855ea9c14b2a72baa8dc76676c32700e0ae8e630/build_O2_fullCI_slc9/pretty.html).

--- a/Utilities/DataSampling/include/DataSampling/Dispatcher.h
+++ b/Utilities/DataSampling/include/DataSampling/Dispatcher.h
@@ -65,7 +65,6 @@ class Dispatcher : public framework::Task
 
  private:
   DataSamplingHeader prepareDataSamplingHeader(const DataSamplingPolicy& policy, header::DataHeader const& original);
-  header::Stack extractAdditionalHeaders(const char* inputHeaderStack) const;
   void reportStats(monitoring::Monitoring& monitoring) const;
   void send(framework::DataAllocator& dataAllocator, const framework::DataRef& inputData, const framework::Output& output) const;
 

--- a/Utilities/DataSampling/src/Dispatcher.cxx
+++ b/Utilities/DataSampling/src/Dispatcher.cxx
@@ -29,6 +29,7 @@
 
 #include <Configuration/ConfigurationInterface.h>
 #include <Configuration/ConfigurationFactory.h>
+#include <stdexcept>
 
 using namespace o2::configuration;
 using namespace o2::monitoring;
@@ -77,6 +78,42 @@ void Dispatcher::init(InitContext& ctx)
   mDeviceID.runtimeInit(spec.id.substr(0, DataSamplingHeader::deviceIDTypeSize).c_str());
 }
 
+header::Stack extractAdditionalHeaders(const char* inputHeaderStack)
+{
+  std::array<header::BaseHeader const*, 5> headers;
+  int count = 0;
+  const auto* first = header::BaseHeader::get(reinterpret_cast<const std::byte*>(inputHeaderStack));
+  for (const auto* current = first; current != nullptr; current = current->next()) {
+    if (current->description != header::DataHeader::sHeaderType && current->description != DataProcessingHeader::sHeaderType) {
+      headers[count++] = current;
+    }
+  }
+
+  // Poor man runtime pack expansion.
+  switch (count) {
+    case 0:
+      return header::Stack{};
+    case 1:
+      return header::Stack{*headers[0]};
+    case 2:
+      return header::Stack{*headers[0], *headers[1]};
+    case 3:
+      return header::Stack{*headers[0], *headers[1], *headers[2]};
+    case 4:
+      return header::Stack{*headers[0], *headers[1], *headers[2], *headers[3]};
+    case 5:
+      return header::Stack{*headers[0], *headers[1], *headers[2], *headers[3], *headers[4]};
+    case 6:
+      return header::Stack{*headers[0], *headers[1], *headers[2], *headers[3], *headers[4], *headers[5]};
+    case 7:
+      return header::Stack{*headers[0], *headers[1], *headers[2], *headers[3], *headers[4], *headers[5], *headers[6]};
+    case 8:
+      return header::Stack{*headers[0], *headers[1], *headers[2], *headers[3], *headers[4], *headers[5], *headers[6], *headers[7]};
+    default:
+      throw std::runtime_error(fmt::format("Too many headers to copy {}", count));
+  }
+}
+
 void Dispatcher::run(ProcessingContext& ctx)
 {
   // todo: consider matching (and deciding) in completion policy to save some time
@@ -106,7 +143,7 @@ void Dispatcher::run(ProcessingContext& ctx)
             // so that custom data-dependent headers are passed forward,
             // and we add a DataSamplingHeader.
             header::Stack headerStack{
-              std::move(extractAdditionalHeaders(part.header)),
+              extractAdditionalHeaders(part.header),
               dsheader};
             const auto* partInputHeader = DataRefUtils::getHeader<header::DataHeader*>(part);
 
@@ -154,21 +191,6 @@ DataSamplingHeader Dispatcher::prepareDataSamplingHeader(const DataSamplingPolic
     policy.getTotalEvaluatedMessages(),
     mDeviceID,
     original};
-}
-
-header::Stack Dispatcher::extractAdditionalHeaders(const char* inputHeaderStack) const
-{
-  header::Stack headerStack;
-
-  const auto* first = header::BaseHeader::get(reinterpret_cast<const std::byte*>(inputHeaderStack));
-  for (const auto* current = first; current != nullptr; current = current->next()) {
-    if (current->description != header::DataHeader::sHeaderType &&
-        current->description != DataProcessingHeader::sHeaderType) {
-      headerStack = std::move(header::Stack{std::move(headerStack), *current});
-    }
-  }
-
-  return headerStack;
 }
 
 void Dispatcher::send(DataAllocator& dataAllocator, const DataRef& inputData, const Output& output) const

